File: rust/src/generic/single_field.rs
use std::fmt;
use crate::generic::i_field::IField;
use crate::generic::i_math::IMath;
use crate::generic::i_ordered::IOrdered;
use crate::generic::i_trigonometric::ITrigonometric;
pub struct SingleField {
    pub f: f32,
    pub print_short: bool,
}

impl SingleField {
    //pub static mut FCOUNT: i32 = 0;

    pub fn new(f: f32) -> Self {
        SingleField { f, print_short: true }
    }

    pub fn copy(&self) -> SingleField {
        SingleField::new(self.f)
    }

    /* pub fn new_array(size: usize) -> Vec<SingleField> {
        vec![SingleField::new(0.0); size]
    } */
}


impl IField for SingleField {
    fn a(&self, o: &SingleField) -> SingleField {
        //unsafe { SingleField::FCOUNT += 1; }
        SingleField::new(self.f + o.f)
    }

    fn ae(&mut self, o: &SingleField) {
        //unsafe { SingleField::FCOUNT += 1; }
        self.f += o.f;

    }

    fn s(&self, o: &SingleField) -> SingleField {
        //unsafe { SingleField::FCOUNT += 1; }
        SingleField::new(self.f - o.f)
    }

    fn se(&mut self, o: &SingleField) {
        //unsafe { SingleField::FCOUNT += 1; }
        self.f -= o.f;
    }

    fn m(&self, o: &SingleField) -> SingleField {
        //unsafe { SingleField::FCOUNT += 1; }
        SingleField::new(self.f * o.f)
    }

    fn me(&mut self, o: &SingleField) {
        //unsafe { SingleField::FCOUNT += 1; }
        self.f *= o.f;
    }

    fn d(&self, o: &SingleField) -> SingleField {
        //unsafe { SingleField::FCOUNT += 1; }
        SingleField::new(self.f / o.f)
    }

    fn de(&mut self, o: &SingleField) {
        //unsafe { SingleField::FCOUNT += 1; }
        self.f /= o.f;
    }

    /*fn coerce(&self) -> f64 {
        self.f as f64
    }*/

    fn is_zero(&self) -> bool {
        self.f == 0.0
    }

    fn is_one(&self) -> bool {
        self.f == 1.0
    }

    fn zero(&self) -> SingleField {
        SingleField::new(0.0)
    }

    fn one(&self) -> SingleField {
        SingleField::new(1.0)
    }

    fn copy(&self) -> SingleField {
        SingleField::new(self.f)
    }
}


impl IMath for SingleField {
    fn abs(&self) -> SingleField {
        SingleField::new(self.f.abs())
    }
    fn sqrt(&mut self) {
        self.f = self.f.sqrt();
    }
}

impl IOrdered for SingleField {
    fn lt(&self, o: &SingleField) -> bool {
        self.f < o.f
    }

    fn le(&self, o: &SingleField) -> bool {
        self.f <= o.f
    }

    fn gt(&self, o: &SingleField) -> bool {
        self.f > o.f
    }

    fn ge(&self, o: &SingleField) -> bool {
        self.f >= o.f
    }

    fn eq(&self, o: &SingleField) -> bool {
        self.f == o.f
    }
}

impl ITrigonometric for SingleField {
    fn sin(&mut self) {
        self.f = self.f.sin();
    }

    fn cos(&mut self) {
        self.f = self.f.cos();
    }
}

impl fmt::Display for SingleField {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if self.print_short {
            write!(f, "{:6.2}", self.f)
        } else {
            write!(f, "{}", self.f)
        }
    }
}

File: java/generic/SingleField.java
package generic;
public class SingleField implements IField<SingleField>,
		ITrigonometric<SingleField>, IMath<SingleField>,
		IOrdered<SingleField>, ICopiable<SingleField> {
	float f;


	//public static int fCount;

	SingleField(float f) {
		this.f = f;
	}

	public SingleField copy() {
		return new SingleField(f);
	}

	public SingleField[] newArray(int size) {
		return new SingleField[size];
	}

	public SingleField a(SingleField o) {
		//fCount++;
		if (o == null)
			return new SingleField(f);
		else
			return new SingleField(f + o.f);
	}

	public void ae(SingleField o) {
		//fCount++;
		if (o != null)
			f += o.f;
	}

	public SingleField s(SingleField o) {
		//fCount++;
		if (o != null)
			return new SingleField(f - o.f);
		else
			return new SingleField(f);
	}

	public void se(SingleField o) {
		//fCount++;
		if (o != null)
			f -= o.f;
	}

	public SingleField m(SingleField o) {
		//fCount++;
		if (o != null)
			return new SingleField(f * o.f);
		else
		return new SingleField(0);
	}

	public void me(SingleField o) {
		//fCount++;
		if (o != null)
			f *= o.f;
		else
			f = 0;
	}

	public SingleField d(SingleField o) {
		//fCount++;
		if (o != null && o.f != 0)
			return new SingleField(f / o.f);
		else
			return new SingleField(0);
	}

	public void de(SingleField o) {
		//fCount++;
		if (o != null && o.f != 0)
			f /= o.f;
		else
			f = 0;
	}


	public SingleField coerce(int i) {
		return new SingleField(i);
	}

	public SingleField coerce(double d) {
		return new SingleField((float) d);
	}

	public double coerce() {
		return f;
	}

	public void sqrt() {
		f = (float) Math.sqrt(f);
	}

	public SingleField sin() {
		return new SingleField((float) Math.sin(f));
	}

	public SingleField cos() {
		return new SingleField((float) Math.cos(f));
	}

 	public String toString() {
			return Float.toString(f);
	} 

	public SingleField newInstance() {
		return new SingleField(0);
	}

	public int intValue() {
		return (int) f;
	}

	public boolean isZero() {
		return f == 0.0;
	}

	public boolean isOne() {
		return f == 1.0;
	}

	public SingleField zero() {
		return new SingleField(0);
	}

	public SingleField one() {
		return new SingleField(1);
	}


	public void abs() {
		if (f < 0)
			f = -f;
	}

	public boolean lt(SingleField o) {
		if (f < o.f)
			return true;
		return false;
	}

	public boolean le(SingleField o) {
		if (f <= o.f)
			return true;
		return false;
	}

	public boolean gt(SingleField o) {
		if (f > o.f)
			return true;
		return false;
	}

	public boolean ge(SingleField o) {
		if (f >= o.f)
			return true;
		return false;
	}

	public boolean eq(SingleField o) {
		if (f == o.f)
			return true;
		return false;
	}
}
